pkg/prometheus/promcfg.go

```diff
@@ -586,6 +613,177 @@ func (cg *configGenerator) generatePodMonitorConfig(
+			})
+		}
+
+		// Set based label matching. We have to map the valid relations
+		// `In`, `NotIn`, `Exists`, and `DoesNotExist`, into relabeling rules.
+		for _, exp := range m.Spec.Targets.Ingress.Selector.MatchExpressions {
+			switch exp.Operator {
+			case metav1.LabelSelectorOpIn:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "keep"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: strings.Join(exp.Values, "|")},
+				})
+			case metav1.LabelSelectorOpNotIn:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "drop"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: strings.Join(exp.Values, "|")},
+				})
+			case metav1.LabelSelectorOpExists:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "keep"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: ".+"},
+				})
+			case metav1.LabelSelectorOpDoesNotExist:
+				relabelings = append(relabelings, yaml.MapSlice{
+					{Key: "action", Value: "drop"},
+					{Key: "source_labels", Value: []string{"__meta_kubernetes_ingress_label_" + sanitizeLabelName(exp.Key)}},
+					{Key: "regex", Value: ".+"},
+				})
+			}
+		}
+
+		if version.Major == 1 && version.Minor < 7 {
+			if apiserverConfig != nil {
+				level.Info(cg.logger).Log("msg", "custom apiserver config is set but it will not take effect because prometheus version is < 1.7")
+			}
+			cfg = append(cfg, cg.generateK8SSDConfig(nil, nil, nil, kubernetesSDRoleIngress))
+		} else {
+			selectedNamespaces := getNamespacesFromNamespaceSelector(&m.Spec.Targets.Ingress.NamespaceSelector, m.Namespace, ignoreNamespaceSelectors)
+			cfg = append(cfg, cg.generateK8SSDConfig(selectedNamespaces, apiserverConfig, basicAuthSecrets, kubernetesSDRoleIngress))
+		}
+
+		// Relabelings for blackbox exporter.
+		relabelings = append(relabelings, []yaml.MapSlice{
+			{
+				{Key: "source_labels", Value: "[__address__]"},
+				{Key: "target_label", Value: "__param_target"},
+			},
+			{
+				{Key: "source_labels", Value: "[__param_target]"},
+				{Key: "target_label", Value: "instance"},
+			},
+			{
+				{Key: "target_label", Value: "__address__"},
+				{Key: "replacement", Value: m.Spec.BlackboxExporter.URL},
+			},
+		}...)
+
+		// Relabelings for ingress SD.
+		relabelings = append(relabelings, []yaml.MapSlice{
+			{
+				{Key: "source_labels", Value: "[__meta_kubernetes_ingress_scheme, __address__, __meta_kubernetes_ingress_path]"},
+				{Key: "separator", Value: ";"},
+				{Key: "regex", Value: "(.+);(.+);(.+)"},
+				{Key: "target_label", Value: "__param_target"},
+				{Key: "replacement", Value: "${1}://${2}${3}"},
+				{Key: "action", Value: "replace"},
```

Comment:
Shouldn't this happen between lines 742 and 743?

Reply:
Why?

Reply:
I'm not sure, but hasn't `__address__` here the value `m.Spec.BlackboxExporter.URL` set above?